### PR TITLE
TSK-1499: Fixed bug of also cleaning completed tasks with parentProcessId null/empty

### PR DESCRIPTION
**Pull request by tge20**
---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [x] Commit message format → TSK-XXX: Your commit message.
- [x] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [x] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [x] PR fulfills the ticket
- [x] Edge cases and unwanted side effects are tested
- [x] Readability

### CODE REVIEW

**Comment by tge20** [comment]
https://sonarcloud.io/project/issues?branch=TSK-1499&id=tge20_taskana&resolved=false&types=CODE_SMELL

**Comment by tge20** [comment]
https://sonarcloud.io/dashboard?branch=TSK-1499&id=tge20_taskana

File: history/taskana-simplehistory-provider/src/test/java/acceptance/jobs/HistoryCleanupJobAccTest.java [hunk 1]
```diff
@@ -394,4 +394,106 @@ void should_DeleteOldHistoryCleanupJobs_When_InitializingSchedule() throws Excep
...
+    taskanaEngine.getConfiguration().setTaskCleanupJobAllCompletedSameParentBusiness(true);
+    HistoryCleanupJob job = new HistoryCleanupJob(taskanaEngine, null, null);
+    job.run();
+
+    assertThat(getHistoryService().createTaskHistoryQuery().count()).isEqualTo(14);
+  }
+
+  @Test
+  @WithAccessId(user = "admin")
+  void should_CleanTaskHistoryEventsWithParentProcessIdEmpty_When_TaskCompleted() throws Exception {
```
Comment (mustaphazorgati) [inline history/taskana-simplehistory-provider/src/test/java/acceptance/jobs/HistoryCleanupJobAccTest.java:451]: This is exactly the same test as above. Please refactor this and use a dynamic test. For examples just search for `@TestFactory`

File: lib/taskana-core/src/test/java/acceptance/jobs/TaskCleanupJobAccTest.java [hunk 2]
```diff
@@ -118,14 +123,59 @@ void should_DeleteOldTaskCleanupJobs_When_InitializingSchedule() throws Exceptio
...
+    taskanaEngine.getConfiguration().setTaskCleanupJobAllCompletedSameParentBusiness(true);
+    taskanaEngine.getConfiguration().setCleanupJobMinimumAge(Duration.ZERO);
+    TaskCleanupJob job = new TaskCleanupJob(taskanaEngine, null, null);
+    job.run();
+
+    List<TaskSummary> tasksAfterCleaning =
+        taskService.createTaskQuery().idIn(taskId1, taskId2).list();
+    assertThat(tasksAfterCleaning).hasSize(1);
+    assertThat(tasksAfterCleaning.get(0).getId()).isEqualTo(taskId2);
+  }
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/jobs/TaskCleanupJobAccTest.java:170]: This test is a duplicate to the test above. Please use dynamic tests to increase maintainability.

File: lib/taskana-core/src/test/java/acceptance/jobs/TaskCleanupJobAccTest.java [hunk 3]
```diff
@@ -118,14 +123,59 @@ void should_DeleteOldTaskCleanupJobs_When_InitializingSchedule() throws Exceptio
...
+
+    taskanaEngine.getConfiguration().setTaskCleanupJobAllCompletedSameParentBusiness(true);
+    taskanaEngine.getConfiguration().setCleanupJobMinimumAge(Duration.ZERO);
+    TaskCleanupJob job = new TaskCleanupJob(taskanaEngine, null, null);
+    job.run();
+
+    List<TaskSummary> tasksAfterCleaning =
+        taskService.createTaskQuery().idIn(taskId1, taskId2).list();
+    assertThat(tasksAfterCleaning).hasSize(1);
+    assertThat(tasksAfterCleaning.get(0).getId()).isEqualTo(taskId2);
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/jobs/TaskCleanupJobAccTest.java:146]:
```suggestion
    assertThat(tasksAfterCleaning).extracting(TaskSummary::getId).containsExactly(taskId2);
```

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

File: history/taskana-simplehistory-provider/src/main/java/pro/taskana/simplehistory/impl/jobs/HistoryCleanupJob.java [hunk 4]
```diff
@@ -113,8 +119,8 @@ public void run() throws TaskanaException {
                 .eventTypeIn(TaskHistoryEventType.CREATED.getName())
                 .list());
 
-        taskIdsToDeleteHistoryEventsFor =
-            filterSameParentBusinessHistoryEventsQualifiedToClean(historyEventCandidatesToClean);
+        taskIdsToDeleteHistoryEventsFor.addAll(
+            filterSameParentBusinessHistoryEventsQualifiedToClean(historyEventCandidatesToClean));
```
[the remark this answers is not in the document]
Comment (tge20) [reply history/taskana-simplehistory-provider/src/main/java/pro/taskana/simplehistory/impl/jobs/HistoryCleanupJob.java:123]: First I add all the events with parentBusinessProcessId NULL or EMPTY, than I check the other events for uncompleted events with the same parentProcess and add those, where none are existing.
But I observed a possible bug: The NULL events should be kicked out from the historyEventCandidatesToClean List such that they may not be added a second time. I will fix it.

**Review by mustaphazorgati — APPROVED** [review]

**Pull request merged**